pkg/interceptors/server/incoming.go

```diff
@@ -0,0 +1,75 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/peer"
+	"net"
+)
+
+type IncomingInterceptor struct {
+	logger *zap.Logger
+}
+
+func NewIncomingInterceptor(logger *zap.Logger) (*IncomingInterceptor, error) {
+	if logger == nil {
+		return nil, fmt.Errorf("logger is required")
+	}
+
+	return &IncomingInterceptor{
+		logger: logger,
+	}, nil
+}
+
+func (i *IncomingInterceptor) logIncomingAddressIfAvailable(ctx context.Context) {
+	if i.logger.Core().Enabled(zap.DebugLevel) {
+		if p, ok := peer.FromContext(ctx); ok {
+			clientAddr := p.Addr.String()
+			var dnsName []string
+			// Attempt to resolve the DNS name
+			host, _, err := net.SplitHostPort(clientAddr)
+			if err == nil {
+				dnsName, err = net.LookupAddr(host)
+				if err != nil || len(dnsName) == 0 {
+					dnsName = []string{"Unknown"}
+				}
+			} else {
+				dnsName = []string{"Unknown"}
+			}
+			i.logger.Debug(
+				fmt.Sprintf("Incoming request from %s (DNS: %s)", clientAddr, dnsName[0]),
```

Comment:
I understand the desire to log incoming node addresses, but could we avoid logging client addresses?

The best solution of course is for clients to anonymize themselves by routing through an intermediary, and/or the protocol itself doing this. But it would be best if the implementation also did its best to minimize unnecessary metadata collection.